### PR TITLE
fix(ios): potential race condition fix

### LINKED ISSUE #2838: 🔥Can't modify config objects after they are in use for FIRDatabaseReferences
## Issue

I'm getting the error displayed in title sometimes after upgrading to v6. Getting it seemingly randomly, unable to intentionally reproduce. Below are screenshots of the error and callstack. They both appear to have some information that might be helpful.

IMG_5323
IMG_5325

---

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# Uncomment the next line to define a global platform for your project
platform :ios, '9.0'

target 'App' do
  # Uncomment the next line if you're using Swift or would like to use dynamic frameworks
  # use_frameworks!

  # Pods for App
  pod 'React', :path => '../node_modules/react-native'
  pod 'yoga', :path => '../node_modules/react-native/ReactCommon/yoga'
  pod 'RNFBApp', :path => '../node_modules/@react-native-firebase/app'
  pod 'RNFBAnalytics', :path => '../node_modules/@react-native-firebase/analytics'
  pod 'RNFBAuth', :path => '../node_modules/@react-native-firebase/auth'
  pod 'RNFBDatabase', :path => '../node_modules/@react-native-firebase/database'
  pod 'RNFBRemoteConfig', :path => '../node_modules/@react-native-firebase/remote-config'
  pod 'RNFBCrashlytics', :path => '../node_modules/@react-native-firebase/crashlytics'
  pod 'Fabric', '~> 1.9.0'
  pod 'Crashlytics', '~> 3.12.0'
  pod 'Folly', :podspec => "../node_modules/react-native/third-party-podspecs/Folly.podspec"

  target 'AppTests' do
    inherit! :search_paths
    # Pods for testing
  end

  post_install do |installer|
    installer.pods_project.targets.each do |target|
      if target.name == "React"
        target.remove_from_project
      end
    end
  end

end
```

#### `AppDelegate.m`:

```objc
#import "AppDelegate.h"

#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>
#import <Firebase.h>
#import <React/RCTPushNotificationManager.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  [FIRApp configure];
  
  NSURL *jsCodeLocation;

  jsCodeLocation = [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];
  
  UIView *backgroundView = [[[NSBundle mainBundle] loadNibNamed:@"LaunchScreen" owner:self options:nil] firstObject];
  RCTRootView *rootView = [[RCTRootView alloc] initWithBundleURL:jsCodeLocation
                                                      moduleName:@"App"
                                               initialProperties:nil
                                                   launchOptions:launchOptions];
  rootView.backgroundColor = [UIColor clearColor];

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = backgroundView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  [backgroundView addSubview:rootView];
  rootView.frame = backgroundView.frame;
  return YES;
}

// Required to register for notifications
 - (void)application:(UIApplication *)application didRegisterUserNotificationSettings:(UIUserNotificationSettings *)notificationSettings
 {
  [RCTPushNotificationManager didRegisterUserNotificationSettings:notificationSettings];
 }
 // Required for the register event.
 - (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken
 {
  [RCTPushNotificationManager didRegisterForRemoteNotificationsWithDeviceToken:deviceToken];
 }
 // Required for the notification event. You must call the completion handler after handling the remote notification.
 - (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo
                                                        fetchCompletionHandler:(void (^)(UIBackgroundFetchResult))completionHandler
 {
   [RCTPushNotificationManager didReceiveRemoteNotification:userInfo fetchCompletionHandler:completionHandler];
 }
 // Required for the registrationError event.
 - (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error
 {
  [RCTPushNotificationManager didFailToRegisterForRemoteNotificationsWithError:error];
 }
 // Required for the localNotification event.
 - (void)application:(UIApplication *)application didReceiveLocalNotification:(UILocalNotification *)notification
 {
  [RCTPushNotificationManager didReceiveLocalNotification:notification];
 }
 
@end
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
 React Native Environment Info:
    System:
      OS: macOS High Sierra 10.13.6
      CPU: (4) x64 Intel(R) Core(TM) i5-2435M CPU @ 2.40GHz
      Memory: 1002.86 MB / 8.00 GB
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 8.7.0 - /usr/local/bin/node
      Yarn: 1.2.1 - /usr/local/bin/yarn
      npm: 5.5.1 - /usr/local/bin/npm
    SDKs:
      iOS SDK:
        Platforms: iOS 12.1, macOS 10.14, tvOS 12.1, watchOS 5.1
    IDEs:
      Android Studio: 3.4 AI-183.6156.11.34.5692245
      Xcode: 10.1/10B61 - /usr/bin/xcodebuild
    npmPackages:
      react: 16.8.3 => 16.8.3 
      react-native: 0.59.3 => 0.59.3 
    npmGlobalPackages:
      create-react-native-app: 1.0.0
      react-native-app-id-fixed: 0.2.1
      react-native-cli: 2.0.1
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [x] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `6.0.3`
- **`Firebase` module(s) you're using that has the issue:**
  - `Realtime Database, maybe RemoteConfig`
- **Are you using `TypeScript`?**
  - `N`
  
</p>
</details>

---

Think `react-native-firebase` is great? Please consider supporting all of the project maintainers and contributors by donating via our [Open Collective](https://opencollective.com/react-native-firebase/donate) where all contributors can submit expenses. [[Learn More]](https://invertase.io/oss/react-native-firebase/contributing/donations-expenses)

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

### PR DESCRIPTION
**Pull request by russellwheatley**
this may fix #2838. As noted in RN docs, the same dispatch queue instance has to be used: https://reactnative.dev/docs/native-modules-ios#threading

There's also some thread locking using the `@synchronized` directive courtesy of @Salakar. We're currently using the  `DISPATCH_QUEUE_SERIAL` macro which at the very least fixes this [problem](https://github.com/invertase/react-native-firebase/issues/2838#issuecomment-572361976). In the future, we might be able to leverage the `@synchronized` directive to make Realtime database calls concurrent using the `DISPATCH_QUEUE_CONCURRENT` macro.  This would help boost performance. 

However, this PR is aimed at the bugs stemming from issue #2838. 

### Checklist

- [ ] Supports `Android`
- [ ] Supports `iOS`
- [ ] `e2e` tests added or updated in packages/\*\*/e2e
- [ ] Flow types updated
- [ ] Typescript types updated

### Release Plan

[CATEGORY][type] [LOCATION] - Message

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Donate via [Open Collective](https://opencollective.com/react-native-firebase/donate)
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter
- 👉 Star this repo on GitHub ⭐️
- 👉 Contribute; see our [contributing guide](/CONTRIBUTING.md)

### CODE REVIEW

File: packages/database/ios/RNFBDatabase/RNFBDatabaseCommon.m [hunk 1]
```diff
@@ -76,33 +87,47 @@ + (void)setDatabaseConfig:(FIRDatabase *)firDatabase
...
-  FIRDatabaseReference *databaseReference = [firebaseDatabase referenceWithPath:path];
+    FIRDatabaseReference *databaseReference = [firebaseDatabase referenceWithPath:path];
 
-  references[key] = databaseReference;
-  return databaseReference;
+    references[key] = databaseReference;
+        
+    return databaseReference;
+    }
 }
```
Comment (Ehesp) [inline packages/database/ios/RNFBDatabase/RNFBDatabaseCommon.m:131]: Formatting is off here

Comment (russellwheatley) [reply packages/database/ios/RNFBDatabase/RNFBDatabaseCommon.m:131]: The formatting is in line with the rest of the code indented within the methods. 5 columns from the left.

Comment (Ehesp) [reply packages/database/ios/RNFBDatabase/RNFBDatabaseCommon.m:131]: Hm, this looks wrong though: 

image

Comment (russellwheatley) [reply packages/database/ios/RNFBDatabase/RNFBDatabaseCommon.m:131]: yeah, just had another look, it's inline with the code above which I have formatted to be two spaces after beginning of method signature. but I didn't take into account the plus/minus sign at the beginning of methods. Will change.

Comment (Salakar) [reply packages/database/ios/RNFBDatabase/RNFBDatabaseCommon.m]: image

Formatting is still off here 👀

Comment (russellwheatley) [reply packages/database/ios/RNFBDatabase/RNFBDatabaseCommon.m]: huh, it is now three columns from the left like all the other code within methods. I don't understand how it is rendering differently in github 🤔:
Screenshot 2020-04-17 at 16 02 19

File: packages/database/ios/RNFBDatabase/RNFBDatabaseCommon.m [hunk 2]
```diff
@@ -186,7 +211,7 @@ + (NSArray *)getCodeAndMessage:(NSError *)error {
       message = [error localizedDescription];
   }
 
-  return @[code, message];
+return @[code, message];
```
Comment (Salakar) [inline packages/database/ios/RNFBDatabase/RNFBDatabaseCommon.m:214]:
```suggestion
  return @[code, message];
```

**Comment by Salakar** [comment]
LGTM!

LGTM GIF

**Pull request merged**